src/lazy/reader.rs

```diff
@@ -275,4 +277,92 @@ mod tests {
         assert_eq!(reader.read_next_element()?, None);
         Ok(())
     }
+
+    fn expand_macro_test(
+        macro_source: &str,
+        encode_macro_fn: impl FnOnce(MacroAddress) -> Vec<u8>,
+        test_fn: impl FnOnce(Reader<AnyEncoding, &[u8]>) -> IonResult<()>,
+    ) -> IonResult<()> {
+        // Because readers do not yet understand encoding directives, we'll pre-calculate the
+        // macro ID that will be assigned. Make an empty encoding context...
+        let context = EncodingContext::empty();
+        // ...and see how many macros it contains. This will change as development continues.
+        let macro_address = context.macro_table.len();
+        let opcode_byte = u8::try_from(macro_address).unwrap();
+        // Using that ID, encode a binary stream containing an invocation of the new macro.
+        // This function must add an IVM and the encoded e-expression ID, followed by any number
+        // of arguments that matches the provided signature.
+        let binary_ion = encode_macro_fn(opcode_byte as usize);
+        // Construct a reader for the encoded data.
+        let mut reader = Reader::new(AnyEncoding, binary_ion.as_slice())?;
+        // Register the template definition, getting the same ID we used earlier.
+        let actual_address = reader.register_template(macro_source)?;
+        assert_eq!(macro_address, actual_address, "Assigned macro address did not match expected address.");
+        // Use the provided test function to confirm that the data expands to the expected stream.
+        test_fn(reader)
+    }
+
+    #[test]
+    fn expand_binary_template_macro() -> IonResult<()> {
+        let macro_source = "(macro seventeen () 17)";
+        let encode_macro_fn = |address| vec![0xE0, 0x01, 0x01, 0xEA, address as u8];
+        expand_macro_test(macro_source, encode_macro_fn, |mut reader| {
+            assert_eq!(reader.expect_next()?.read()?.expect_i64()?, 17);
+            Ok(())
+        })
+    }
```

Comment:
🗺️ The tests in this file are doing actual expansion of binary encoded e-expressions.